module/helper/renderer/src/webgl/material/mod.rs

```diff
@@ -0,0 +1,252 @@
+    /// The material uses a mask based on an alpha cutoff value.
+    Mask,
+    /// The material uses standard alpha blending.
+    Blend
+  }
+
+  /// Represents the cull mode for the material
+  #[ derive( Default, Clone, Copy, PartialEq, Eq, Debug ) ]
+  pub enum CullMode
+  {
+    /// Cull front face
+    Front,
+    /// Cull back face
+    #[ default ]
+    Back,
+    /// Cull back and front face
+    FrontAndBack
+  }
+
+  /// Represents the depth function
+  #[ derive( Default, Clone, Copy, PartialEq, Eq, Debug ) ]
+  pub enum DepthFunc
+  {
+    /// Never pass
+    Never,
+    /// Pass if the incoming value is less than the depth buffer value
+    #[ default ]
+    Less,
+    /// Pass if the incoming value equals the depth buffer value
+    Equal,
+    /// Pass if the incoming value is less than or equal to the depth buffer value
+    LEqual,
+    /// Pass if the incoming value is greater than the depth buffer value
+    Greater,
+    /// Pass if the incoming value is not equal to the depth buffer value
+    NotEqual,
+    /// Pass if the incoming value is greater than or equal to the depth buffer value
+    GEqual,
+    /// Always pass
+    Always
+  }
+
+  /// Stores information about a texture used by the material, including the texture itself and its UV coordinates.
+  ///
+  /// You may have several attibutes for the UV coordinates in the shader:
+  /// `
+  /// layout( location = 0 ) in vec2 uv_0;
+  /// layout( location = 1 ) in vec2 uv_1;
+  /// `
+  /// uv_position will defines which UV to use
+  #[ derive( Default, Clone, Debug ) ]
+  pub struct TextureInfo
+  {
+    /// The texture object.
+    pub texture : Rc< RefCell< Texture > >,
+    /// The UV coordinate set index to use for this texture.
+    pub uv_position : u32
+  }
+
+  impl TextureInfo
+  {
+    /// Uploads the texture data to the GPU.
+    pub fn upload( &self, gl : &gl::WebGl2RenderingContext )
+    {
+      self.texture.borrow().upload( gl );
+    }
+
+    /// Binds the texture to a texture unit.
+    pub fn bind( &self, gl : &gl::WebGl2RenderingContext )
+    {
+      self.texture.borrow().bind( gl );
+    }
+  }
+
+  /// A trait representin a generic material
+  pub trait Material : std::any::Any + Debug
+  {
+    /// Returns the unique identifier of the material.
+    fn get_id( &self ) -> uuid::Uuid;
+
+    /// Returns a human-readable name for the material (for debugging/editor).
+    fn get_name( &self ) -> Option< &str >
+    {
+      None
+    }
+
+    /// Signal for updating material uniforms
+    fn is_need_update( &self ) -> bool;
+
+    /// Returns answer need use IBL for current material instance or not
+    fn is_need_use_ibl( &self ) -> bool
+    {
+      false
+    }
+
+    /// Can or not use this material IBL
+    fn can_use_ibl( &self ) -> bool
+    {
+      false
+    }
+
+    /// Returns [`ProgramInfo`] with shader locations and used [`ShaderProgram`]
+    fn get_program_info( &self, gl : &GL, program : &gl::WebGlProgram ) -> ProgramInfo;
```

Comment:
Why do we need to pass "program" as parameter and why do we return full type and not a reference?